make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+        if too_few_unique_sequences:
+            return True
+
+        num_unique_gapped_seqs = get_number_of_unique_gapped_sequences(alignment)
+
+        # this is an assert as if it happens, something very wrong with the two previous functions is happening
+        assert num_unique_nongapped_seqs <= num_unique_gapped_seqs
+
+        alignment_has_ambiguity = num_unique_nongapped_seqs < num_unique_gapped_seqs
+        if alignment_has_ambiguity:
+            # TODO: fix alignment by deduplicating nongapped seqs and realigning them
+            # TODO: the annoying thing is that this deduplicated nongapped alignment will differ from the input alignment
+            # TODO: but I think it can be an ok solution
+            return True
+
+        return False
+
+    @staticmethod
+    def _get_nesting_level(parent_node: Optional[RecursiveTreeNode]) -> int:
+        is_root = parent_node is None
+        if is_root:
+            nesting_level = 0
+        else:
+            need_to_go_down_one_level = isinstance(parent_node, MultiClusterNode)
+            if need_to_go_down_one_level:
+                nesting_level = parent_node.nesting_level + 1
+            else:
+                nesting_level = parent_node.nesting_level
+        return nesting_level
+    #####################################################################################################
+
+    #####################################################################################################
+    #  interval methods
+    @staticmethod
+    def _infer_if_has_single_interval(all_intervals: Intervals, match_intervals: Intervals,
+                                      nesting_level: int, max_nesting: int,
+                                      alignment: MSA, min_match_length: int) -> bool:
+        single_match_interval = (len(all_intervals) == 1) and (all_intervals[0] in match_intervals)
+        if single_match_interval:
+            return True
+
+        max_nesting_level_reached = nesting_level == max_nesting
+        if max_nesting_level_reached:
+            return True
+
+        small_variant_site = alignment.get_alignment_length() < min_match_length
+        if small_variant_site:
+            return True
+
+        return False
+
+    @staticmethod
+    def _get_all_and_match_intervals(alignment: MSA, min_match_length: int) -> Tuple[Intervals, Intervals]:
+        consensus = get_consensus_from_MSA(alignment)
+        interval_partitioner = IntervalPartitioner(consensus, min_match_length, alignment)
+        (
+            match_intervals,
+            non_match_intervals,
+            all_intervals,
+        ) = interval_partitioner.get_intervals()
+        return all_intervals, match_intervals
+
+    @staticmethod
+    def _partition_alignment_into_interval_subalignments(alignment: MSA, all_intervals: Intervals) -> List[MSA]:
+        return [alignment[:, interval.start: interval.stop + 1] for interval in all_intervals]
```

Comment:
I think this should just be done in `NodeFactory.build()` and not be a function, but leave it to your choice

Reply:
I will leave as is because is much easier to test, i.e. it is much easier to mock a function call than mocking several calls to the slice method of biopython's `MultipleSeqAlignment` class... Also, is better documentation than trying to read what the list comprehension does

Reply:
i see and agree